naga/src/diagnostic_filter.rs

```diff
@@ -0,0 +1,137 @@
+//! [`DiagnosticFilter`]s and supporting functionality.
+
+use crate::Handle;
+#[cfg(feature = "wgsl-in")]
+use crate::Span;
+#[cfg(feature = "wgsl-in")]
+use indexmap::IndexMap;
+
+/// A severity set on a [`DiagnosticFilter`].
+///
+/// <https://www.w3.org/TR/WGSL/#diagnostic-severity>
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub enum Severity {
+    Off,
+    Info,
+    Warning,
+    Error,
+}
+
+impl Severity {
+    const ERROR: &'static str = "error";
+    const WARNING: &'static str = "warning";
+    const INFO: &'static str = "info";
+    const OFF: &'static str = "off";
+
+    /// Convert from a sentinel word in WGSL into its associated [`Severity`], if possible.
+    pub fn from_ident(s: &str) -> Option<Self> {
+        Some(match s {
+            Self::ERROR => Self::Error,
+            Self::WARNING => Self::Warning,
+            Self::INFO => Self::Info,
+            Self::OFF => Self::Off,
+            _ => return None,
+        })
+    }
+}
+
+/// A filterable triggering rule in a [`DiagnosticFilter`].
+///
+/// <https://www.w3.org/TR/WGSL/#filterable-triggering-rules>
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub enum FilterableTriggeringRule {
+    DerivativeUniformity,
+}
+
+impl FilterableTriggeringRule {
+    const DERIVATIVE_UNIFORMITY: &'static str = "derivative_uniformity";
+
+    /// Convert from a sentinel word in WGSL into its associated [`FilterableTriggeringRule`], if possible.
+    pub fn from_ident(s: &str) -> Option<Self> {
+        Some(match s {
+            Self::DERIVATIVE_UNIFORMITY => Self::DerivativeUniformity,
+            _ => return None,
+        })
+    }
+
+    /// Maps this [`FilterableTriggeringRule`] into the sentinel word associated with it in WGSL.
+    pub const fn to_ident(self) -> &'static str {
+        match self {
+            Self::DerivativeUniformity => Self::DERIVATIVE_UNIFORMITY,
+        }
+    }
+
+    #[cfg(feature = "wgsl-in")]
+    pub(crate) const fn tracking_issue_num(self) -> u16 {
+        match self {
+            FilterableTriggeringRule::DerivativeUniformity => 5320,
+        }
+    }
+}
+
+/// A filter that modifies how diagnostics are emitted for shaders.
+///
+/// <https://www.w3.org/TR/WGSL/#diagnostic-filter>
+#[derive(Clone, Debug)]
+pub struct DiagnosticFilter {
+    pub new_severity: Severity,
+    pub triggering_rule: FilterableTriggeringRule,
+}
+
+/// A map of diagnostic filters to their severity and first occurrence's span.
+///
+/// Intended for front ends' first step into storing parsed [`DiagnosticFilter`]s.
+#[derive(Clone, Debug, Default)]
+#[cfg(feature = "wgsl-in")]
+pub(crate) struct DiagnosticFilterMap(IndexMap<FilterableTriggeringRule, (Severity, Span)>);
```

Comment:
This PR only populates this data structure so, I can't tell if its shape is what we need when consuming it. Could you move it into the follow-up that implements the full behavior/logic for diagnostics?

Reply:
That's a reasonable request. 🙂 I'll move this into <https://github.com/gfx-rs/wgpu/pull/6148>.

Reply:
This is almost not true, funnily enough. This code _would_ be used for conflict detection, except we return an error on the first recognized triggering rule anyway. 😅

Moved to #6148 with 2ea2cd04 and its children.